downloadDirectory/src/main/java/org/reactome/release/downloadDirectory/GenerateGOAnnotationFile.java

```diff
@@ -0,0 +1,413 @@
+        }
+        return false;
+    }
+
+    // This checks if the protein in question has UniProt as a Reference Database, and if the species has a crossReference. If not, it's information will not be generated for the GOA file.
+    // TODO: Rule check
+    private static boolean validProtein(GKInstance referenceEntityInst, GKInstance speciesInst) throws Exception {
+        if (referenceEntityInst != null && speciesInst != null) {
+            GKInstance referenceDatabaseInst = (GKInstance) referenceEntityInst.getAttributeValue(ReactomeJavaConstants.referenceDatabase);
+            if (referenceDatabaseInst != null && referenceDatabaseInst.getDisplayName().equals("UniProt") && speciesInst.getAttributeValue(ReactomeJavaConstants.crossReference) != null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // Finds most recent modification date for a GOA line. This is a bit of a moving target since GOA lines can be generated convergently for
+    // each type of GO annotation. Depending on if it is looking at the individual protein or whole reaction level, the date attribute
+    // may not be the most recent. If it is found that the goaLine was generated earlier but that a more recent modification date exists based on the
+    // entity that is currently being checked, then it will just update that date value in the hash associated with the line. (Yes, this is weird).
+    private static Integer assignDateForGOALine(GKInstance entityInst, String goaLine) throws Exception {
+        int instanceDate;
+        Collection<GKInstance> modifiedInstances = entityInst.getAttributeValuesList(ReactomeJavaConstants.modified);
+        if (modifiedInstances.size() > 0) {
+            List<GKInstance> modifiedInstancesList = new ArrayList<>(modifiedInstances);
+            GKInstance mostRecentModifiedInst = modifiedInstancesList.get(modifiedInstancesList.size() - 1);
+            instanceDate = Integer.valueOf(mostRecentModifiedInst.getAttributeValue(ReactomeJavaConstants.dateTime).toString().split(" ")[0].replaceAll("-", ""));
+        } else {
+            GKInstance createdInst = (GKInstance) entityInst.getAttributeValue(ReactomeJavaConstants.created);
+            instanceDate = Integer.valueOf(createdInst.getAttributeValue(ReactomeJavaConstants.dateTime).toString().split(" ")[0].replaceAll("-", ""));
+        }
+
+        // Stores date in global hash that allows date value to be updated if a more recent date was found.
+        if (dates.get(goaLine) == null) {
+            dates.put(goaLine, instanceDate);
+        } else {
+            if (instanceDate > dates.get(goaLine)) {
+                dates.put(goaLine, instanceDate);
+            }
+        }
+        return instanceDate;
+    }
+
+    // Generates the third column of the GOA line by checking attributes of the referenceEntity instance.
+    private static String getSecondaryIdentifier(GKInstance referenceEntityInst) throws Exception {
```

Comment:
This method uses similar logic to get the values and check them.  Maybe it could be done like this:
```
// Generates the third column of the GOA line by checking attributes of the referenceEntity instance.
private static String getSecondaryIdentifier(GKInstance referenceEntityInst) {
    List<String> attributes = Arrays.asList(
        ReactomeJavaConstants.secondaryIdentifier, 
	ReactomeJavaConstants.geneName,
	ReactomeJavaConstants.identifier
    );
    return attributes
	.stream()
	.map(attribute -> getAttributeValue(referenceEntityInst, attribute))
	.filter(Objects::nonNull)
	.findFirst()
	.orElse("");
}

private static String getAttributeValue(GKInstance instance, String attribute) {
	try {
		return instance.getAttributeValue(attribute).toString();
	} catch (Exception e) {
		return "";
	}
}
```

Reply:
That adds a lot more writing to the function, so I think for this suggestion I will stick with my implementation